File: leet-code/2022/0002-add-two-numbers.cpp
// https://leetcode.com/problems/add-two-numbers

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        return addTwoNumbers(l1, l2, 0);
    }

    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2, int remaining) {
        int sum = remaining;
        if (l1) {
            sum += l1->val + (l2 ? l2->val : 0);
            l1->val = sum % 10;
            l1->next = addTwoNumbers(l1->next, l2 ? l2->next : nullptr, sum/10);
            return l1;
        }
        if (l2) {
            sum += l2->val;
            l2->val = sum % 10;
            l2->next = addTwoNumbers(l2->next, nullptr, sum/10);
            return l2;
        }
        return remaining ? new ListNode(remaining) : nullptr;

    }
};

class FirstUglySolution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        return addTwoNumbers(l1, l2, 0);
    }

    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2, int remaining) {
        int sum = remaining;
        if (l1 && l2) {
            sum += l1->val + l2->val;
            l1->val = sum % 10;
            l1->next = addTwoNumbers(l1->next, l2->next, sum/10);
            return l1;
        }

        auto longest = l1 ? l1 : l2;
        if (!longest) {
            return remaining ? new ListNode(remaining) : nullptr;
        }

        sum += longest->val;
        longest->val = sum % 10;
        if (!longest->next && sum/10) {
            longest->next = new ListNode(remaining);
        } else {
            longest->next = addTwoNumbers(longest->next, nullptr, sum/10);
        }
        return longest;
    }
};

/*
[9,9,9,9,9,9,9]
[9,9,9,9]
*/

File: leet-code/2022/0002-add-two-numbers.rs
// https://leetcode.com/problems/add-two-numbers

// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
//
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn add_two_numbers(l1: Option<Box<ListNode>>, l2: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        Self::add_two_numbers_with_remaining(l1, l2, 0)
    }

    fn add_two_numbers_with_remaining(
        l1: Option<Box<ListNode>>,
        l2: Option<Box<ListNode>>,
        remaining: i32
    ) -> Option<Box<ListNode>> {
        let mut sum = remaining;
        match (l1, l2) {
            // both lists are None
            (None, None) => {
                if remaining > 0 {
                    return Some(Box::new(ListNode::new(remaining)));
                }
                None
            }
            // only 1 list is None
            (None, Some(mut list)) | (Some(mut list), None) => {
                sum += list.val;
                list.val = sum % 10;
                list.next = Self::add_two_numbers_with_remaining(list.next, None, sum/10);
                Some(list)
            },
            // none is None
            (Some(mut list), Some(l2)) => {
                sum += list.val + l2.val;
                list.val = sum % 10;
                list.next = Self::add_two_numbers_with_remaining(list.next, l2.next, sum/10);
                Some(list)
            },
        }
    }
}
